pkg/controller/support/credentials/cloudauth/gcr/gcr_test.go

```diff
@@ -1,102 +1,252 @@
+	"github.com/dominodatalab/hephaestus/pkg/controller/support/credentials/cloudauth/cloudauthtest"
 )
 
-var envCredentials = "GOOGLE_APPLICATION_CREDENTIALS"
+var defaultTestingErr = errors.New("default error message")
 
-func TestRegister(t *testing.T) {
-	if os.Getenv(envCredentials) == "" {
-		t.Skip("Skipping, gcp not setup")
+func TestAuthenticate(t *testing.T) {
+	defaultCtx := context.Background()
+
+	// Canceled context for testing failed do request
+	canceledCtx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	// logger for comparing expected vs actual logging.
+	observerCore, observedLogs := observer.New(zap.DebugLevel)
+	log := zapr.NewLogger(zap.New(observerCore))
+
+	// test server
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "New Server, %s", r.Proto)
+	}))
+
+	t.Cleanup(ts.Close)
+
+	// test server IP and port information for error logging
+	tsAddress := ts.Listener.Addr()
+
+	// expected errors
+	invalidUrlErr := errors.New(fmt.Sprintf("invalid GCR URL test-ts should match %s", gcrRegex))
+	gcrTokenAccessErr := errors.New("unable to access GCR token from oauth: default error message")
+	gcrRegistryErr := errors.New("GCR registry \"https://hi-docker.pkg.dev\" is unusable: default error message")
+	ctxCanceledErr := errors.New(fmt.Sprintf("request to access GCR reqistry token failed with Error: Get \"http://oauth2accesstoken:***@%s?client_id=forge&service=serviceName\": context canceled", tsAddress.String()))
+	non200StatusErr := errors.New("failed to obtain token, received unexpected response code: 400\nresponse: \"nope\"")
+	noResTokenErr := errors.New("no GCR token in bearer response:\n{\"token\":\"\",\"access_token\":\"\",\"refresh_token\":\"\"}")
+
+	for _, tt := range []struct {
+		name               string
+		serverName         string
+		ctx                context.Context
+		roundTripper       roundTripFunc
+		tokenShouldErr     bool
+		authConfig         *types.AuthConfig
+		loginChallenger    cloudauth.LoginChallenger
+		expectedLogMessage string
+		expectedError      error
+	}{
+		{
+			name:               "invalid-url",
+			serverName:         "test-ts",
+			ctx:                defaultCtx,
+			roundTripper:       nil,
+			authConfig:         nil,
+			loginChallenger:    defaultChallengeLoginServer,
+			expectedLogMessage: invalidUrlErr.Error(),
+			expectedError:      invalidUrlErr,
+		},
+		{
+			name:               "oauth2-error",
+			serverName:         "gcr.io",
+			ctx:                defaultCtx,
+			tokenShouldErr:     true,
+			loginChallenger:    defaultChallengeLoginServer,
+			expectedLogMessage: gcrTokenAccessErr.Error(),
+			expectedError:      gcrTokenAccessErr,
+		},
+		{
+			name:       "failed-challenge-ts-error",
+			serverName: "hi-docker.pkg.dev",
+			ctx:        defaultCtx,
+			loginChallenger: cloudauthtest.FakeChallengeLoginServer(
+				"serviceName",
+				"realmName",
+				defaultTestingErr),
```

Comment:
```suggestion
			loginChallenger: cloudauthtest.FakeChallengeLoginServer(
				"serviceName",
				"realmName",
				defaultTestingErr,
			),
```

I would reformat all of these blocks in this test table.